File: my-chattester/client/src/GetContext.py
# GetContext.py
# -*- coding: utf-8 -*-
import ast
import sys
import io
import os

sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

class FunctionContextExtractorFromFile:
    def __init__(self, file_path, target_function_name):
        self.file_path = file_path
        self.target_function_name = target_function_name
        self.source_code = ""
        self.ast_tree = None
        self.target_function_node = None
        self.context = {
            'imports': [],
            'globals': [],
            'class_definitions': {},
            'function_definitions': {},
            'target_function': None,
            'dependencies': []
        }

    def load_source(self):
        with open(self.file_path, 'r', encoding='utf-8') as f:
            self.source_code = f.read()
        self.ast_tree = ast.parse(self.source_code)

    def find_target_function(self):
        for node in ast.walk(self.ast_tree):
            if isinstance(node, ast.FunctionDef) and node.name == self.target_function_name:
                self.target_function_node = node
                break

    def extract_imports(self):
        for node in ast.walk(self.ast_tree):
            if isinstance(node, (ast.Import, ast.ImportFrom)):
                self.context['imports'].append(ast.unparse(node))

    def extract_globals(self):
        for node in self.ast_tree.body:
            if isinstance(node, ast.Assign) and isinstance(node.targets[0], ast.Name):
                self.context['globals'].append({
                    'name': node.targets[0].id,
                    'value': ast.unparse(node.value)
                })

    def extract_class_and_function_definitions(self):
        for node in self.ast_tree.body:
            if isinstance(node, ast.ClassDef):
                methods = {}
                for item in node.body:
                    if isinstance(item, ast.FunctionDef):
                        methods[item.name] = ast.unparse(item)
                self.context['class_definitions'][node.name] = {
                    'signature': ast.unparse(node),
                    'methods': methods
                }
            elif isinstance(node, ast.FunctionDef):
                self.context['function_definitions'][node.name] = ast.unparse(node)

    def analyze_dependencies(self):
        if self.target_function_node is None:
            return
        for node in ast.walk(self.target_function_node):
            if isinstance(node, ast.Call):
                if isinstance(node.func, ast.Name):
                    self.context['dependencies'].append(node.func.id)
                elif isinstance(node.func, ast.Attribute):
                    self.context['dependencies'].append(node.func.attr)

    def get_type_context(self):
        if self.target_function_node is None:
            return {}
        parameters = []
        for arg in self.target_function_node.args.args:
            param_info = {
                'name': arg.arg,
                'type': ast.unparse(arg.annotation) if arg.annotation else 'Any',
                'default': None
            }
            parameters.append(param_info)
        return_type = ast.unparse(self.target_function_node.returns) if self.target_function_node.returns else 'Any'

        return {
            'function_name': self.target_function_node.name,
            'parameters': parameters,
            'return_type': return_type
        }

    def build_context(self):
        self.load_source()
        self.find_target_function()
        self.extract_imports()
        self.extract_globals()
        self.extract_class_and_function_definitions()
        self.analyze_dependencies()

        function_def_str = ast.unparse(self.target_function_node) if self.target_function_node else ''

        return {
            'type_context': self.get_type_context(),
            'dependency_context': {
                'imports': self.context['imports'],
                'globals': self.context['globals'],
                'dependencies': self.context['dependencies']
            },
            'class_context': {},  # 可拓展：查找所属类
            'function_definition': function_def_str,
            'related_definitions': {
                'classes': self.context['class_definitions'],
                'functions': self.context['function_definitions']
            }
        }


def format_context(context):
    output = []

    type_ctx = context['type_context']
    output.append("=== 函数基本信息 ===")
    output.append(f"函数名称: {type_ctx.get('function_name', '')}")
    output.append(f"返回类型: {type_ctx.get('return_type', 'Any')}")
    output.append("\n参数列表:")
    for param in type_ctx.get('parameters', []):
        param_str = f"- {param['name']}: {param['type']}"
        output.append(param_str)

    dep_ctx = context['dependency_context']
    output.append("\n=== 依赖信息 ===")
    output.append("\n导入语句:")
    for imp in dep_ctx['imports']:
        output.append(f"- {imp}")

    output.append("\n全局变量:")
    for global_var in dep_ctx['globals']:
        output.append(f"- {global_var['name']}: {global_var['value']}")

    output.append("\n依赖的函数/方法:")
    for dep in dep_ctx['dependencies']:
        output.append(f"- {dep}")

    output.append("\n=== 相关定义 ===")
    output.append("\n其他类定义:")
    for class_name, class_def in context['related_definitions']['classes'].items():
        output.append(f"\n类 {class_name}:\n{class_def['signature']}")

    output.append("\n其他函数定义:")
    for func_name, func_def in context['related_definitions']['functions'].items():
        output.append(f"\n函数 {func_name}:\n{func_def}")

    output.append("\n=== 完整函数定义 ===")
    output.append(context['function_definition'])

    return "\n".join(output)


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python GetContext.py <file_path> <target_function>")
        sys.exit(1)

    file_path = sys.argv[1]
    target_function_name = sys.argv[2]

    extractor = FunctionContextExtractorFromFile(file_path, target_function_name)
    context = extractor.build_context()
    formatted = format_context(context)
    print(formatted)

File: my-chattester/client/src/GetContext.js
const Parser = require('node-tree-sitter');
const Python = require('tree-sitter-python');
const fs = require('fs');
const path = require('path');
const { spawnSync } = require('child_process');

const parser = new Parser();
parser.setLanguage(Python);

function extractContext(sourceCode, functionName) {
    const tree = parser.parse(sourceCode);
    const root = tree.rootNode;

    const context = {
        imports: [],
        globals: [],
        class_definitions: {},
        function_definitions: {},
        target_function: null,
        dependencies: [],
        type_context: {
            function_name: functionName,
            parameters: [],
            return_type: 'Any'
        }
    };

    function extractAnnotation(node) {
        if (!node) return 'Any';
        return sourceCode.slice(node.startIndex, node.endIndex);
    }

    for (const child of root.namedChildren) {
        if (child.type === 'import_statement' || child.type === 'import_from_statement') {
            context.imports.push(sourceCode.slice(child.startIndex, child.endIndex));

        } else if (child.type === 'expression_statement') {
            const expr = child.namedChildren[0];
            if (expr && expr.type === 'assignment') {
                const debugSource = sourceCode.slice(expr.startIndex, expr.endIndex);
                console.log("\n[DEBUG] 🔍 分析 assignment 节点源码:\n" + debugSource);

                const named = expr.namedChildren;
                console.log(`[DEBUG] 🧩 namedChildren 长度: ${named.length}`);
                named.forEach((n, i) => {
                    console.log(`[DEBUG]   ➤ 第 ${i + 1} 个子节点类型: ${n.type}, 文本: ${sourceCode.slice(n.startIndex, n.endIndex)}`);
                });

                // 简单单变量赋值
                if (named.length === 2 && named[0].type === 'identifier') {
                    const name = named[0].text;
                    const value = sourceCode.slice(named[1].startIndex, named[1].endIndex);
                    console.log(`[DEBUG] ✅ 添加全局变量: ${name} = ${value}`);
                    context.globals.push({ name, value });
                }

                // 多变量赋值
                else if (named.length === 2 && named[0].type === 'expression_list') {
                    const leftList = named[0].namedChildren;
                    const rightList = named[1].namedChildren;
                    for (let i = 0; i < leftList.length; i++) {
                        const left = leftList[i];
                        const right = rightList[i];
                        if (left.type === 'identifier') {
                            const name = left.text;
                            const value = right ? sourceCode.slice(right.startIndex, right.endIndex) : 'undefined';
                            console.log(`[DEBUG] ✅ 添加多变量: ${name} = ${value}`);
                            context.globals.push({ name, value });
                        }
                    }
                } else {
                    console.warn(`[DEBUG] ⚠️ 未处理的 assignment 结构，跳过：${debugSource}`);
                }
            }
        }


        else if (child.type === 'class_definition') {
            const className = child.childForFieldName('name').text;
            const methods = {};
            for (const item of child.namedChildren) {
                if (item.type === 'function_definition') {
                    const methodName = item.childForFieldName('name').text;
                    methods[methodName] = sourceCode.slice(item.startIndex, item.endIndex);
                }
            }
            context.class_definitions[className] = {
                signature: sourceCode.slice(child.startIndex, child.endIndex),
                methods
            };
        } else if (child.type === 'function_definition') {
            const funcName = child.childForFieldName('name').text;
            const funcCode = sourceCode.slice(child.startIndex, child.endIndex);
            context.function_definitions[funcName] = funcCode;

            if (funcName === functionName) {
                context.target_function = child;

                const params = child.childForFieldName('parameters');
                for (const param of params.namedChildren) {
                    if (param.type === 'identifier') {
                        context.type_context.parameters.push({
                            name: param.text,
                            type: 'Any'
                        });
                    } else if (param.type === 'default_parameter') {
                        const nameNode = param.child(0);
                        const name = nameNode ? nameNode.text : '[未知参数]';
                        const annotation = nameNode && nameNode.namedChildren.length > 1
                            ? nameNode.namedChildren[1]
                            : null;
                        context.type_context.parameters.push({
                            name,
                            type: extractAnnotation(annotation)
                        });
                    } else if (param.type === 'typed_parameter') {
                        const nameNode = param.childForFieldName('name');
                        const annotation = param.childForFieldName('type');
                        const name = nameNode ? nameNode.text : '[未知参数]';
                        context.type_context.parameters.push({
                            name,
                            type: extractAnnotation(annotation)
                        });
                    } else {
                        context.type_context.parameters.push({
                            name: '[未知参数]',
                            type: 'Any'
                        });
                    }
                }

                const returnTypeNode = child.childForFieldName('return_type');
                if (returnTypeNode) {
                    context.type_context.return_type = extractAnnotation(returnTypeNode);
                }

                const calls = child.descendantsOfType('call');
                for (const call of calls) {
                    const funcNode = call.child(0);
                    if (funcNode) {
                        context.dependencies.push(funcNode.text);
                    }
                }
            }
        }
    }

    return context;
}

function formatContext(context) {
    const output = [];

    const typeCtx = context.type_context;
    output.push("=== 函数基本信息 ===");
    output.push(`函数名称: ${typeCtx.function_name}`);
    output.push(`返回类型: ${typeCtx.return_type}`);
    output.push("\n参数列表:");
    for (const param of typeCtx.parameters) {
        output.push(`- ${param.name}: ${param.type}`);
    }

    output.push("\n=== 依赖信息 ===");
    output.push("\n导入语句:");
    for (const i of context.imports) {
        output.push(`- ${i}`);
    }

    output.push("\n全局变量:");
    for (const g of context.globals) {
        output.push(`- ${g.name}: ${g.value}`);
    }

    output.push("\n依赖的函数/方法:");
    for (const d of context.dependencies) {
        output.push(`- ${d}`);
    }

    output.push("\n=== 相关定义 ===");
    output.push("\n其他类定义:");
    for (const [name, cls] of Object.entries(context.class_definitions)) {
        output.push(`\n类 ${name}:\n${cls.signature}`);
    }

    output.push("\n其他函数定义:");
    for (const [name, def] of Object.entries(context.function_definitions)) {
        if (name !== typeCtx.function_name) {
            output.push(`\n函数 ${name}:\n${def}`);
        }
    }

    output.push("\n=== 完整函数定义 ===");
    output.push(context.function_definitions[typeCtx.function_name] || "未找到目标函数");

    return output.join('\n');
}

// function runFunctionIOTest(context) {
//     if (!context.function_definitions[context.type_context.function_name]) {
//         console.log("\n🚫 未找到目标函数，无法执行测试。");
//         return;
//     }

//     console.log("\n=== 🧪 输入输出示例测试 ===");

//     const testCases = [
//         [12, true, "yes"],
//         [12, false, "yes"],
//         [12, false, "no"],
//         [7, true, "yes"],
//         [3, false, "no"]
//     ];

//     const funcCode = context.function_definitions[context.type_context.function_name];
//     const runnerPath = path.join(__dirname, 'temp_run.py');
//     const runnerCode = `
// ${funcCode}

// import sys
// import json

// if __name__ == "__main__":
//     args = json.loads(sys.argv[1])
//     print(${context.type_context.function_name}(*args))
// `;

//     fs.writeFileSync(runnerPath, runnerCode, 'utf-8');

//     for (const args of testCases) {
//         const input = JSON.stringify(args);
//         const result = spawnSync('python', [runnerPath, input], {
//             encoding: 'utf-8'
//         });

//         const output = result.stdout.trim();
//         console.log(`输入 ${JSON.stringify(args)} => 输出: ${output}`);
//     }

//     fs.unlinkSync(runnerPath);
// }

// === CLI 执行部分 ===
if (require.main === module) {
    const [,, filePath, functionName] = process.argv;

    if (!filePath || !functionName) {
        console.log("用法: node GetContext.js <python_file_path> <target_function>");
        process.exit(1);
    }

    const code = fs.readFileSync(filePath, 'utf-8');
    const context = extractContext(code, functionName);
    const output = formatContext(context);
    console.log(output);

    // runFunctionIOTest(context);
}

module.exports = {
    extractContext,
    formatContext
};